wideq/core.py

```diff
@@ -59,6 +62,32 @@ def get_wideq_logger() -> logging.Logger:
 LOGGER = get_wideq_logger()
 
 
+@lru_cache(maxsize=5)
```

Comment:
I'm a little confused—why do we need an LRU cache here? It seems like a single session would suffice, right? And doesn't `lru_cache` disambiguate based on the function arguments (of which there are none)?

Reply:
A `Session` instance in requests greatly benefits from persistence and re-use.  It allows multiple requests to reuse the shared connection pool and will often allow subsequent requests to utilize the same open TCP connection (eliminating the overhead of opening a TLS-encrypted TCP connection).

This was my way of ensuring that `get_retry_session` always returned the same session instance (since there are no arguments, every call will resolve to the same cache item and return that).  I added size 5 in case there was ever a reason in the future to instantiate another few sessions.

For instance, maybe it would be best to have separate sessions for the login/token requests and one for the requests to the service (each preloaded with the appropriate headers), or maybe you/someone adds a callback feature that hits another service (e.g., hass REST API) that wants a separate set of retry-able status codes/backoff/etc.

In that case, all you have to do is add some parameters to the function and the lru_cache will ensure maximum session reuse (which is a good thing).

I could have replaced all that by adding a line

```
SESSION = get_retry_session()
```

at the module level and gotten rid of the cache.  But what I've written here has the same functional effect with a slightly more extensible 'design'.  But it may also be less clear.

If you'd like to see something else here instead, just let me know.

Reply:
Got it; thanks for the detailed explanation!

I think I would like to stick with the `SESSION` global for this… to me, it makes it clearer what's going on: that we're using a global, module-level session for all the functions that seem to be "pure" in this module.

I am also still nervous about a couple of things:

- It's not 100% clear to me from the Requests docs whether it is safe to have a single, global, shared, never-closed Session for all requests in the lifetime of a process. We have no need to persist cookies, but we will now do that between all requests, for example. And the docs have examples showing how to *close* a session after use, which we will now never do—even if the client application closes/deletes all the `wideq.core.Session` objects. Which is sort of an unpleasant use of global state, if it matters. Maybe it would be better to pass in `Session` objects to these methods?
- That aforementioned technique would also let the client customize, for example, the number of retries. I don't think "retry effectively indefinitely" is a good default—it could cause code to hang seemingly forever if the service is down, for example, which may be fine for a background process but not for something that is supposed to be interactive. But of course it would be nice to support extremely large numbers of retries if the client code needs it.

Finally, one last concern: are we sure that these status codes indicate that a request can be idempotently retried? For example, a service might emit an error even when it took an action—it seems like a potential problem that we might then transparently/automatically try again, possibly taking the action twice. Is that something we should worry about?

Reply:
> * It's not 100% clear to me from the Requests docs whether it is safe to have a single, global, shared, never-closed Session for all requests in the lifetime of a process.

What do you mean by "safe" here?  I do this regularly.  The underlying connections in the pool may close, but the pool handles reopening them if necessary.  I don't think the API passes cookies around, and I've been running this on my own hassio instance for about a week with only one warning (that being that the first request failed and the retry succeeded instead of outright crashing).

> Finally, one last concern: are we sure that these status codes indicate that a request can be idempotently retried?

No.  That said, the only thing I've ever observed is the connection failing to establish (resolved by `connect=NUM_RETRIES`).  If their use of status codes is "standard", it would be safe to treat 502, 503, and 504 as retry-able codes (Bad Gateway, Service Unavailable, and Gateway Timeout, respectively).  500 is often indicative of a longer-term issue (it's often thrown when there's an uncaught exception in many different web frameworks, for example).

I'll change the status codes to 502, 503, and 504.

> * I don't think "retry effectively indefinitely" is a good default—it could cause code to hang seemingly forever if the service is down, for example, which may be fine for a background process but not for something that is supposed to be interactive.

Fair enough.  These changes are what I implemented for myself running in hass (in which wideq is a background poller), but I can change this to whatever you'd like.

> For example, a service might emit an error even when it took an action—it seems like a potential problem that we might then transparently/automatically try again, possibly taking the action twice. Is that something we should worry about?

If this happens when the service throws a 5xx, then it's a service bug.  I've seen it happen (mostly my code), but it shouldn't.  Most of the time for 502 and 503, the request hasn't even hit the service (it catches on the gateway or reverse proxy and gets thrown back before ever hitting application code).  500 and 504 would be more at risk for that.  I'm happy dropping retries for those codes if you'd rather, but being resilient to this really is LG's responsibility.

Reply:
> I think I would like to stick with the `SESSION` global for this… to me, it makes it clearer what's going on: that we're using a global, module-level session for all the functions that seem to be "pure" in this module.

Done.  lru_cache removed.

> * I don't think "retry effectively indefinitely" is a good default

Reduced to 5.  I've yet to observe it need more than 1 retry on my hassio setup.

> Finally, one last concern: are we sure that these status codes indicate that a request can be idempotently retried? For example, a service might emit an error even when it took an action—it seems like a potential problem that we might then transparently/automatically try again, possibly taking the action twice. Is that something we should worry about?

I changed the status codes to 502, 503, and 504 which are unlikely to have reached the server in the first place.

Reply:
Awesome; thank you for the detailed explanations!

I'm going to merge this with a couple of small tweaks. The one medium-sized thing is that I think I'll switch to creating sessions "on the fly," which will result in the smallest possible change w/r/t the current behavior. I'm just nervous about adding more global, mutable state (the SESSION object) that is unavoidable even when you use the "functional" API.

In the future, if it separately seems like a persistent Session would be useful for performance reasons, we can explore that—but the Requests Session object can be associated with a wideq Session object (or similar), which will require a bit more plumbing.